File: modules/remotebackend/unittest_json.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'bundler/setup'
require 'json'
require 'thread'
require 'webrick'
require './unittest'

class DNSBackendHandler < WEBrick::HTTPServlet::AbstractServlet
   def initialize(server, dnsbackend)
     @dnsbackend = dnsbackend
     @semaphore = Mutex.new
     @f = File.open("/tmp/remotebackend.txt.#{$$}","a")
     @f.sync
   end

   def do_POST(req,res)
     req.continue

     return 400, "Bad request" unless req.path == "/dns/endpoint.json"

     tmp = JSON::parse(req.body)
     method = tmp["method"].downcase
     method = "do_#{method}"
     args = tmp["parameters"]

     @f.puts "#{Time.now.to_f} [http/json]: #{({:method=>method,:parameters=>args}).to_json}"    

     @semaphore.synchronize do
       if @dnsbackend.respond_to?(method.to_sym)
          result, log = @dnsbackend.send(method.to_sym, args)
          body = {:result => result, :log => log}
          res.status = 200
          res["Content-Type"] = "application/javascript; charset=utf-8"
          res.body = body.to_json
        else
          res.status = 404
          res["Content-Type"] = "application/javascript; charset=utf-8"
          res.body = ({:result => false, :log => ["Method not found"]}).to_json
        end
        @f.puts "#{Time.now.to_f} [http/json]: #{res.body}"
     end
   end
end

server = WEBrick::HTTPServer.new(
	:Port=>62434,
	:BindAddress=>"localhost",
#	Logger: WEBrick::Log.new("remotebackend-server.log"),
	:AccessLog=>[ [ File.open("remotebackend-access.log", "w"), WEBrick::AccessLog::COMBINED_LOG_FORMAT ] ] 
)

be = Handler.new 
server.mount "/dns", DNSBackendHandler, be
server.mount_proc("/ping"){ |req,resp| resp.body = "pong" }

trap('INT') { server.stop }
trap('TERM') { server.stop }

server.start

File: modules/remotebackend/unittest_json.py
#!/usr/bin/env python

import http.server
import json
from pdns_unittest import Handler


class DNSBackendServer(http.server.HTTPServer):
    def __init__(self, *args, **kwargs):
        self.handler = Handler()
        super().__init__(*args, **kwargs)

    def finish_request(self, request, client_address):
        """Finish one request by instantiating RequestHandlerClass."""
        self.RequestHandlerClass(request, client_address, self, handler=self.handler)


class DNSBackendHandler(http.server.BaseHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        self.handler = kwargs["handler"]
        super().__init__(*args)

    def do_GET(self):
        if self.path == "/ping":
            self.send_response(200)
            self.end_headers()
            self.wfile.write("pong".encode())
            return
        self.send_error(404)

    def do_POST(self):
        if self.path != "/dns/endpoint.json":
            self.send_error(404)
            return

        try:
            length = int(self.headers.get("content-length"))
            message = json.loads(self.rfile.read(length).decode())

            method = "do_" + message["method"].lower()
            args = message["parameters"]
            self.handler.result = False
            self.handler.log = []

            if callable(getattr(self.handler, method, None)):
                getattr(self.handler, method)(**args)
                result = json.dumps({"result": self.handler.result, "log": self.handler.log}).encode()

                self.send_response(200)
                self.send_header("content-type", "text/javascript")
                self.send_header("content-length", len(result))
                self.end_headers()
                self.wfile.write(result)
            else:
                self.send_error(404, message=json.dumps({"error": "No such method"}))
        except BrokenPipeError as e2:
            raise e2
        except Exception as e:
            self.send_error(400, message=str(e))


def main():
    server = DNSBackendServer(("", 62434), DNSBackendHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass


main()
